yoti-sdk-impl/src/main/java/com/yoti/api/client/spi/remote/util/AnchorCertificateUtils.java

```diff
@@ -0,0 +1,111 @@
+import java.util.Enumeration;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+import org.bouncycastle.asn1.ASN1InputStream;
+import org.bouncycastle.asn1.ASN1OctetString;
+import org.bouncycastle.asn1.ASN1Primitive;
+import org.bouncycastle.asn1.ASN1TaggedObject;
+import org.bouncycastle.asn1.DEROctetString;
+import org.bouncycastle.asn1.DLSequence;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.protobuf.ByteString;
+import com.yoti.api.client.spi.remote.proto.AttrProto.Anchor;
+
+public class AnchorCertificateUtils {
+    private static final Logger LOG = LoggerFactory.getLogger(AnchorCertificateUtils.class);
+
+   
+    public static AnchorVerifierSourceData getTypesFromAnchor(Anchor anchor) {
+        Set<String> types = new HashSet<String>(anchor.getOriginServerCertsCount());
+        AnchorType anchorType = AnchorType.UNKNOWN;
+
+        try {
+            CertificateFactory factory = CertificateFactory.getInstance("X.509");
+
+            for (ByteString bs : anchor.getOriginServerCertsList()) {
+                X509Certificate certificate = (X509Certificate) factory.generateCertificate(new ByteArrayInputStream(bs.toByteArray()));
+
+                List<String> extensions = new ArrayList<String>();
+                for (AnchorType type : AnchorType.values()) {
+                    extensions = getListOfStringFromExtension(certificate, type.extensionOid);
+                    if (extensions.size() > 0) {
+                        anchorType = type;
+                        break;
+                    }
+                }
+                types.addAll(extensions);
+            }
+        } catch (Exception e) {
+            LOG.warn("Could not extract anchor type from certificate.", e);
+        }
+
+        return new AnchorVerifierSourceData(types, anchorType);
+    }
+
+    
+    private static List<String> getListOfStringFromExtension(X509Certificate certificate, String extensionValue) throws IOException {
+        List<String> extensionsStrings = new ArrayList<String>();
+
+        byte[] extension = certificate.getExtensionValue(extensionValue);
+        if (extension != null) {
+            // Read the First object
+            ASN1InputStream asn1InputStream = new ASN1InputStream(extension);
+            ASN1Primitive derObject = asn1InputStream.readObject();
+
+            if (derObject != null && derObject instanceof DEROctetString) {
+                DEROctetString derOctetString = (DEROctetString) derObject;
+
+                // Read the sub object which is expected to be a sequence
+                ASN1InputStream derAsn1stream = new ASN1InputStream(derOctetString.getOctets());
+                DLSequence dlSequence = (DLSequence) derAsn1stream.readObject();
+
+                // Enumerate all the objects in the sequence, we expect only one !
+                Enumeration<?> seqEnum = dlSequence.getObjects();
+                while (seqEnum.hasMoreElements()) {
```

Comment:
If we only expect one, maybe we should just us nextElement() and then null check it.
Should we fail if there is more than one?

Reply:
We shouldn't fail and it is actually future proof in case anchors end up having more than one string value per object id, as the ASN.1 format allows it.

It is fair to question why would we add more strings to an object, but I rather have a piece of code that accounts for that possibility than one that might leave out valuable data.